extensions/azurecore/src/account-provider/auths/azureAuth.ts

```diff
@@ -0,0 +1,491 @@
+		const response: TokenResponse = {};
+
+		for (const tenant of azureAccount.properties.tenants) {
+			let cachedTokens = await this.getCachedToken(account.key, resource.id, tenant.id);
+			// Check expiration
+			if (cachedTokens) {
+				const expiresOn = Number(this.memdb.get(`${account.key.accountId}_${tenant.id}_${resource.id}`));
+				const currentTime = new Date().getTime() / 1000;
+
+				if (!Number.isNaN(expiresOn)) {
+					const remainingTime = expiresOn - currentTime;
+					// expired
+					if (remainingTime < 5 * 60) {
+						cachedTokens = undefined;
+					}
+				} else {
+					// No expiration date, assume expired.
+					cachedTokens = undefined;
+					console.log('Assuming expired token');
+				}
+
+			}
+
+			// Refresh
+			if (!cachedTokens) {
+
+				const baseToken = await this.getCachedToken(account.key);
+				if (!baseToken) {
+					return undefined;
+				}
+
+				await this.refreshAccessToken(account.key, baseToken.refreshToken, tenant, resource);
+				cachedTokens = await this.getCachedToken(account.key, resource.id, tenant.id);
+				if (!cachedTokens) {
+					return undefined;
+				}
+			}
+			const { accessToken } = cachedTokens;
+			response[tenant.id] = {
+				token: accessToken.token,
+				key: accessToken.key,
+				tokenType: 'Bearer'
+			};
+		}
+
+		if (azureAccount.properties.subscriptions) {
+			azureAccount.properties.subscriptions.forEach(subscription => {
+				response[subscription.id] = {
+					...response[subscription.tenantId]
+				};
+			});
+		}
+
+		return response;
+	}
+
+	public async clearCredentials(account: azdata.AccountKey): Promise<void> {
+		try {
+			return this.deleteAccountCache(account);
+		} catch (ex) {
+			const msg = localize('azure.cacheErrrorRemove', "Error when removing your account from the cache.");
+			console.error('Error when removing tokens.', ex);
+			throw new Error(msg);
+		}
+	}
+
+	protected toBase64UrlEncoding(base64string: string): string {
+		return base64string.replace(/=/g, '').replace(/\+/g, '-').replace(/\//g, '_'); // Need to use base64url encoding
+	}
+
+	protected async makePostRequest(uri: string, postData: { [key: string]: string }) {
+		const config = {
+			headers: {
+				'Content-Type': 'application/x-www-form-urlencoded'
+			}
+		};
+
+		return axios.post(uri, qs.stringify(postData), config);
+	}
+
+	protected async makeGetRequest(token: string, uri: string): Promise<AxiosResponse<any>> {
+		const config = {
+			headers: {
+				Authorization: `Bearer ${token}`,
+				'Content-Type': 'application/json',
+			},
+		};
+
+		return axios.get(uri, config);
+	}
+
+	protected async getTenants(token: AccessToken): Promise<Tenant[]> {
+		interface TenantResponse { // https://docs.microsoft.com/en-us/rest/api/resources/tenants/list
+			id: string
+			tenantId: string
+			displayName?: string
+			tenantCategory?: string
+		}
+
+		const tenantUri = url.resolve(this.metadata.settings.armResource.endpoint, 'tenants?api-version=2019-11-01');
```

Comment:
It feels like this api version should be a part of the config or something?

Reply:
At the very least this should be a const since you use it in multiple places

Reply:
The API seems to update independently of eachother from what I've seen